### PR TITLE
Stale connection mitigation in blaze-client

### PR DESCRIPTION
**Pull request by rossabaker**
Adds an automated retry policy to blaze-client to account for connections that closed while idling in the pool.

This is less flexible than Ember's, mainly because we don't have access to a `Timer`.  Then again, this is the bare minimum suggested by HTTP specs, and we don't need all that power.  Like Ember's, this is a conservative default that only retries idempotent requests.

This, like Ember's default strategy, may interact surprisingly with an explicit retry middleware as the retry policies compose.

### CODE REVIEW

File: blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala [hunk 1]
```diff
@@ -86,49 +88,53 @@ object BlazeClient {
...
-                new SocketException(s"HTTP connection closed: ${key}")
-              }
-              .map { (response: Resource[F, Response[F]]) =>
-                response.flatMap(r =>
-                  Resource.make(F.pure(r))(_ => manager.release(next.connection))
-                )
-              }
+        def loop(retriesRemaining: Int): F[Resource[F, Response[F]]] =
+          borrow
+            .use { next =>
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala:93]: Ayo, WTF, scalafmt?

File: blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala [hunk 2]
```diff
@@ -86,49 +88,53 @@ object BlazeClient {
...
+                    F.racePair(gate.get *> res, responseHeaderTimeoutF)
+                      .flatMap[Resource[F, Response[F]]] {
+                        case Left((r, fiber)) => fiber.cancel.as(r)
+                        case Right((fiber, t)) => fiber.cancel >> F.raiseError(t)
+                      }
+                  }
+                case _ => res
+              }
+            }
+            .recoverWith {
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala:132]: Concern: are we certain this only borrows one connection at a time?  This is the golden rule of the connection pool, and previous naive flatmaps have deadlocked the connection pool.

Comment (rossabaker) [reply blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala:132]: I gave the client a max of one connection per key, so I think we're good.

**Comment by rossabaker** [comment]
Recommend viewing this one without whitespace diffs.  It's smaller than it looks.

**Review by ChristopherDavenport — APPROVED** [review]

File: blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClientBuilder.scala [hunk 3]
```diff
@@ -368,6 +428,7 @@ object BlazeClientBuilder {
       asynchronousChannelGroup = None,
       channelOptions = ChannelOptions(Vector.empty),
       customDnsResolver = None,
+      retries = 2,
```
Comment (danicheg) [inline blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClientBuilder.scala:431]: Not a blocker at all, I'm just interested in why '2' was chosen here?

**Review by danicheg — APPROVED** [review]
LGTM.

File: blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClientBuilder.scala (hunk 3 above)
Comment (rossabaker) [reply blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClientBuilder.scala:431]: 1. It matches Ember.
2. I think Ember matches an RFC, but damn if I can find it in there now.

Comment (danicheg) [reply blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClientBuilder.scala:431]: Got it, thanks!

**Pull request merged**